File: cc/test/mixintest2.cc
#include <iostream>

using namespace std;

template <typename T, typename Superclass> class ConstrainedCollection: public Superclass {
public:
    virtual void push_back(T x) {
        if (Superclass::size() > 5) {
            throw 42;
        }
        cout << "fuege Element hinzu: " << x << endl;
        Superclass::push_back(x);
    }
};


#include <vector>




int main() {
    try {
        ConstrainedCollection<int, vector<int> > arr;
        arr.push_back(1);
        arr.push_back(2);
        arr.push_back(3);
        arr.push_back(4);
        arr.push_back(5);
        arr.push_back(6);
        arr.push_back(7);
        arr.push_back(8);
        arr.push_back(9);
    }
    catch (int err) {
        cout <<  "Fehler: " << err << endl;
    }
}

File: cc/test/mixintest2.rb
module ConstrainedCollection

  def push(x)
    if size() > 5
      raise "Collection wird zu gross!"
    end
    print "fuege Element hinzu: #{x}\n";
    super(x)
  end

end




arr=[1,2]
arr.push 3
=> [1, 2, 3]
irb(main):004:0> arr.extend ConstrainedCollection
=> [1, 2, 3]
irb(main):005:0> arr.push 4
fuege Element hinzu: 4
=> [1, 2, 3, 4]
irb(main):006:0> arr.push 5
fuege Element hinzu: 5
=> [1, 2, 3, 4, 5]
irb(main):007:0> arr.push 6
fuege Element hinzu: 6
=> [1, 2, 3, 4, 5, 6]
irb(main):008:0> arr.push 7
